File: Perm/JAVA/Perm.java
import java.util.Scanner;

public class Perm {

    public static void perm (Object[] list, int k, int m) {
        if (k == m) {
            for (int i = 0; i <= m; i++) {
                System.out.print(list[i] + " ");
            }
            System.out.println();
        } else {
            for (int i = k; i <= m; i++) {
                swap(list, k, i);
                perm(list, k + 1, m);
                swap(list, k, i);
            }
        }
    }

    public static void swap(Object[] list, int k, int m) {
        Object t;
        t = list[k];
        list[k] = list[m];
        list[m] = t;
    }

    public static void main(String[] args) {
        String[] list = new String[100];
        int i = 0;

        Scanner sc = new Scanner(System.in);
        String s = sc.nextLine();
        String str[] = s.split(" ");
        for (String num : str) {
            list[i++] = num;
        }

        System.out.println();
        perm(list, 0, i - 1);
    }
}

File: Perm/C_Cpp/perm.cpp
#include<iostream>
using namespace std;

void swap(char *a, char *b)
{
    int t;
    t = *a;
    *a = *b;
    *b = t;
}

void perm(char *list, int k, int m)
{
    if (k == m) {
        for (int i = 0; i <= m; i++) {
            cout << *(list + i) << " ";
        }
        cout << endl;
    } else {
        for (int i = k; i <= m; i++) {
            swap(list + k, list + i);
            perm(list, k + 1, m);
            swap(list + k, list + i);
        }
    }
}

int main()
{
    char list_a[] = {'4'};
    char list_b[] = {'a', 'b', 'c', 'd'};

    perm(list_a, 0, 0);
    cout << endl;
    perm(list_b, 0, 3);

    return 0;
}
